flashlight/fl/examples/RnnClassification.cpp

```diff
@@ -0,0 +1,260 @@
+
+class RnnClassifier : public Container {
+ public:
+  explicit RnnClassifier(unsigned num_classes, unsigned vocab_size, int hidden_size = 256, unsigned num_layers=2)
+      : embed(Embedding(hidden_size, vocab_size)),
+        rnn(RNN(hidden_size, hidden_size, num_layers, RnnMode::GRU, 0 /* Dropout */)),
+        linear(Linear(hidden_size, num_classes)),
+        logsoftmax(0) {
+    std::cout << "Creating a RNN Classifier with vocab size: " << vocab_size << " and num classes: " << num_classes << std::endl;
+    add(embed);
+    add(rnn);
+    add(linear);
+    add(LogSoftmax());
+  }
+
+  std::vector<Variable> forward(const std::vector<Variable>& inputs) override { return rnn(inputs); }
+
+  std::tuple<Variable, Variable, Variable>
+  forward(const Variable& input, const Variable& h, const Variable& c) {
+    FL_VLOG(fl::INFO) << "Embed input shape: " << input.dims();
+    const unsigned numchars = input.dims(1);
+    Variable ho, co; // hidden and carry output
+    Variable output = embed(input); // output should be hs x bs x ts : [ 𝑋𝑖𝑛, 𝑁, 𝑇 ]
+    // The input to the RNN is expected to be of shape [ 𝑋𝑖𝑛, 𝑁, 𝑇 ] where 𝑋𝑖𝑛 is the hidden size, 𝑁 is the batch size and 𝑇 is the sequence length.
+    FL_VLOG(fl::INFO) << "RNN input (expect H,N,T): " << output.dims();
+    std::tie(output, ho, co) = rnn(output, h, c);
+    // The output of the RNN should be of shape [ 𝑋𝑜𝑢𝑡, 𝑁, 𝑇], with 𝑋𝑜𝑢𝑡 to be the hidden_size (unidirectional RNN)
+    // Truncate BPTT
+    ho.setCalcGrad(false);
+    co.setCalcGrad(false);
+    output = linear(output);    
+    output = logsoftmax(output);
+    output = output(af::span, af::span, input.array().dims(1) - 1); // check with reviewers : default FL rnn direction is guarantee to be Left2right meaning output to be the last col?
+    return std::make_tuple(output, ho, co);
+  }
+
+  std::string prettyString() const override { return "RnnClassifer"; }
+
+  // Inference on the given input: returns the category index
+  unsigned infer(const std::string& inputstring, const Variable& h, const Variable& c) {
+    af::array ia = ClassDataset::lineToTensor(inputstring);
+    Variable output, ho, co;
+    std::tie(output, ho, co) = forward(noGrad(ia), h,c);
+    af::array max_value, prediction;
+    af::max(max_value, prediction, output.array(), 0);
+    unsigned classid = prediction.scalar<unsigned>();
+    FL_VLOG(fl::INFO) << inputstring << " : " << ClassDataset::ID2label[classid];
+    return classid;
+  }
+
+  // Predict the category of the given input, compared with the expected label and print the result
+  bool unittest(const std::string& input, const std::string& expectedLabel) {
+    Variable output, h, c;
+    auto p = ClassDataset::ID2label[infer(input,h,c)];
+    const std::string s = (p == expectedLabel ? "✓ " : "✗ ");
+    std::cout << "input: " << std::setw(20) << input << "\t expected: " << expectedLabel << "\t prediction: " <<  p << "\t" << s << std::endl;
+  }
+
+ private:
+  Embedding embed;
```

Comment:
please use "_" at the end of each private field

Reply:
done